entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -290,19 +296,115 @@ private void doUpdate(RequestContext requestContext, EntityUpdateRequest request
           EntityQueryConverter.convertToAttributeValue(setAttribute.getValue()).build();
       String jsonValue = DocStoreJsonFormat.printer().print(attributeValue);
 
+      Map<Key, Map<String, Document>> toUpdate = new HashMap<>();
+
       for (String entityId : request.getEntityIdsList()) {
         SingleValueKey key =
             new SingleValueKey(requestContext.getTenantId().orElseThrow(), entityId);
         // TODO better error reporting once doc store exposes the,
-        if (!entitiesCollection.updateSubDoc(key, subDocPath, new JSONDocument(jsonValue))) {
-          LOG.warn(
-              "Failed to update entity {}, subDocPath {}, with new doc {}.",
-              key,
-              subDocPath,
-              jsonValue);
+        if (toUpdate.containsValue(key)) {
+          toUpdate.get(key).put(subDocPath, new JSONDocument(jsonValue));
+        } else {
+          Map<String, Document> subDocument = new HashMap<>();
+          subDocument.put(subDocPath, new JSONDocument(jsonValue));
+          toUpdate.put(key, subDocument);
         }
       }
+      try {
+        entitiesCollection.bulkUpdateSubDocs(toUpdate);
+      } catch (Exception e) {
+        LOG.warn(
+            "Failed to update entities, subDocPath {}, with new doc {}.", subDocPath, jsonValue);
+      }
+    }
+  }
+
+  @Override
+  public void bulkUpdate(
+      BulkEntityUpdateRequest request, StreamObserver<ResultSetChunk> responseObserver) {
+    // Validations
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    Optional<String> tenantId = requestContext.getTenantId();
+    if (tenantId.isEmpty()) {
+      responseObserver.onError(new ServiceException("Tenant id is missing in the request."));
+      return;
+    }
+    if (StringUtils.isEmpty(request.getEntityType())) {
+      responseObserver.onError(new ServiceException("Entity type is missing in the request."));
+      return;
+    }
+    if (request.getEntitiesCount() == 0) {
+      responseObserver.onError(new ServiceException("Entity IDs are missing in the request."));
+    }
+    try {
+      bulkDoUpdate(requestContext, request);
+      // Finally return the selections
+      List<String> entityIdsList = new ArrayList<>();
```

Comment:
```suggestion
      Set<String> entityIdsList = request.getEntitiesMap().keySet();
```

And, then can get rid of the for loop completely